pkg/ipcache/metadata.go

```diff
@@ -361,10 +361,26 @@ func (ipc *IPCache) doInjectLabels(ctx context.Context, modifiedPrefixes []netip
 				break
 			}
 
+			// If the old entry is tainted, and we are attempting to overwrite it,
+			// then we want to keep its old source around in case we eventually
+			// stop managing it and need to restore the old source
+			var newOverwrittenLegacySource source.Source
+			switch {
+			case oldID.exclusivelyOwnedByLegacyAPI():
+				// This is the first time we have associated metadata for a
+				// tainted entry. Store the old (legacy) source
+				newOverwrittenLegacySource = oldID.Source
+			case oldID.ownedByLegacyAndMetadataAPI():
+				// The entry is tainted, but has already been updated at
+				// least once by the metadata API. Retain the legacy source as is.
+				newOverwrittenLegacySource = oldID.overwrittenLegacySource
+			}
+
 			// We can safely skip the ipcache upsert if the entry matches with
 			// the entry in the metadata cache exactly.
 			// Note that checking ID alone is insufficient, see GH-24502.
 			if oldID.ID == newID.ID && prefixInfo.Source() == oldID.Source &&
+				oldID.overwrittenLegacySource == newOverwrittenLegacySource &&
```

Comment:
Should this use `.equals()`? I suspect so (and other versions may have been a bug). But this code is ticklish.

Reply:
This put me down the rabbit hole:

The only fields not checked by the old if statement here are are `modifiedByLegacyAPI` (formerly known as `tainted`) and `shadowed`.

- `modifiedByLegacyAPI` by definition will never be changed by an metadata upsert (similar to `createdFromMetadata` the value is sticky and can never flip once touched by the legacy API). If it cannot change, then it will always compare in equals and can be omitted from the check
- `shadowed` is not something we really know the new value of here. It depends on the existence of a covering endpoint IP, or (I believe) a cilium_host IP. At the moment, we do not have endpoint IPs via the metadata API, but cilium_host is. I'm honestly not sure if the code here was correct or what the correct handling regarding shadowing even should be. 

My assumption is the whole shadowing logic is not needed if everything was handled via the metadata API, since the metadata API itself already deals with the cases the old "shadowed" logic is trying handle.

Thus, I _believe_ that if the old entry was shadowed, then the new one will be shadowed too (since any changes to covering endpoint IP legacy upserts cannot happen concurrently with `InjectLabels`). This means that comparing it here is futile, and that the check is correct as is.

I'll leave as is for now. But I'll take another look at this with fresh eyes tomorrow

Reply:
Also, unrelated side-note, this if statement here should check for `entryExists` before reading `oldID`

Reply:
I decided not to use `equals` for the reasons mentioned above. Please let me know what you think